File: sdks/conformance/bindings/typescript/src/case-spec.ts
// In-memory representation of one conformance YAML file. Field names
// mirror the YAML/openapi.yaml snake_case convention so the loader can
// pass keys through verbatim.

export interface CaseSpec {
  name: string;
  description?: string;
  runner_config: RunnerConfig;
  handlers: HandlerSpec[];
  server_script: ScriptEntry[];
  shutdown_after_ms?: number;
  expectations: Expectations;
}

export interface RunnerConfig {
  runner_id?: string;
  runner_id_prefix?: string;
  capabilities?: string[];
  tags?: string[];
  max_inflight?: number;
  api_key?: string;
  bearer_token?: string;
  poll_timeout_ms?: number;
  renew_interval_ms?: number;
  drain_timeout_ms?: number;
  poll_retry_delay_ms?: number;
  capacity_backoff_ms?: number;
  max_consecutive_poll_conflicts?: number;
}

export type HandlerBehavior = 'noop' | 'throw' | 'sleep' | 'log' | 'stream_logs';

export interface HandlerSpec {
  job_key: string;
  is_default?: boolean;
  schedule?: string;
  behavior: HandlerBehavior;
  error_message?: string;
  duration_ms?: number;
  level?: 'trace' | 'debug' | 'info' | 'warn' | 'error';
  message?: string;
  count?: number;
  interval_ms?: number;
}

export interface ScriptEntry {
  on: string;
  match_count?: number;
  respond: RespondSpec;
}

export interface RespondSpec {
  status: number;
  body?: unknown;
  delay_ms?: number;
  headers?: Record<string, string>;
}

export interface Expectations {
  duration_max_ms?: number;
  http: HttpExpectation[];
}

export interface HttpExpectation {
  method: string;
  path: string;
  exact_count?: number;
  min_count?: number;
  max_count?: number;
  headers?: Record<string, string>;
  body_match?: unknown;
}

File: sdks/java/conformance-tests/src/test/java/io/croniq/runner/conformance/CaseSpec.java
package io.croniq.runner.conformance;

import java.util.List;
import java.util.Map;

/**
 * In-memory representation of one conformance case YAML. Mirrors the
 * structure declared in {@code sdks/conformance/schema/case-schema.json}.
 *
 * <p>Field names use camelCase; the YAML keys are snake_case and translated
 * by {@link CaseLoader}. Nested {@code body} payloads and {@code body_match}
 * specs keep their loose Map/List form so {@link BodyMatcher} can do JSON
 * subset matching without an intermediate object model.
 */
public record CaseSpec(
        String name,
        String description,
        RunnerConfig runnerConfig,
        List<HandlerSpec> handlers,
        List<ScriptEntry> serverScript,
        Integer shutdownAfterMs,
        Expectations expectations) {

    public record RunnerConfig(
            String runnerId,
            String runnerIdPrefix,
            List<String> capabilities,
            List<String> tags,
            Integer maxInflight,
            String apiKey,
            String bearerToken,
            Integer pollTimeoutMs,
            Integer renewIntervalMs,
            Integer drainTimeoutMs,
            Integer pollRetryDelayMs,
            Integer capacityBackoffMs,
            Integer maxConsecutivePollConflicts,
            Integer maxConsecutiveAuthFailures) {}

    public record HandlerSpec(
            String jobKey,
            Boolean isDefault,
            String schedule,
            String behavior,
            String errorMessage,
            Integer durationMs,
            String level,
            String message,
            Integer count,
            Integer intervalMs) {}

    public record ScriptEntry(String on, Integer matchCount, Respond respond) {

        public record Respond(int status, Object body, Map<String, String> headers, Integer delayMs) {}
    }

    public record Expectations(Integer durationMaxMs, List<HttpExpectation> http) {

        public record HttpExpectation(
                String method,
                String path,
                Integer exactCount,
                Integer minCount,
                Integer maxCount,
                Map<String, String> headers,
                Object bodyMatch) {}
    }
}
